### PR TITLE
[MRG] Datacite 4.2 Support

### PR DESCRIPTION
**Pull request by tmorrell**
Add support for the DataCite 4.2 metadata schema.  This uses the official DataCite schema and tests against their examples.  I've had to make some modifications to both the schema and examples, which are under discussion here: https://github.com/datacite/schema/pull/55

I haven't finished fixing the issues with the xml-> json conversion for 4.2 metadata.

Closes #32

### CODE REVIEW

**Comment by tmorrell** [comment]
All my new code for adding 4.2 now passes the tests.  All the rest of the failures are in the other tests.  I'll try to adjust versions to get travis to be happy.

**Comment by tmorrell** [comment]
I think this is ready for review and merging.  All the httpretty tests were broken with , so to get those to pass I've swapped in responses instead.  It was pretty much a 1-1 replacement, except for the timeout test which I commented out.

**Comment by lnielsen** [comment]
Perfect. I'm including the review in our next sprint which is starting on Monday 22th. If I manage I'll try to get through it before.

File: datacite/jsonutils.py [hunk 1]
```diff
@@ -18,14 +18,13 @@
 
 def validator_factory(schema_filename):
```
Comment (ppanero) [inline datacite/jsonutils.py]: Minor detail, but these changes are leaving all the import from ``os`` unused. They are not showed in this diff, but are the following:
``` python 
from os.path import abspath, basename, dirname, split
```
Maybe is a good idea to remove them to keep it clean.

File: datacite/jsonutils.py [hunk 2]
```diff
@@ -18,14 +18,13 @@
...
-    schema_name = basename(schema_filename)
-
-    with open(schema_filename) as fp:
-        schema_json = json.load(fp)
+    with open(schema_filename, 'r') as fp:
+        schema = json.load(fp)
 
     resolver = jsonschema.RefResolver(
-        'file://'+'/'.join(split(schema_dir)) + '/', schema_name
+    'file:{}'.format(schema_filename),
```
Comment (ppanero) [inline datacite/jsonutils.py]: Shouldn't this be in the same line and tabbed?
``` python
resolver = jsonschema.RefResolver(
        'file:{}'.format(schema_filename), schema
)
```

File: datacite/schema42.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of DataCite.
+#
+# Copyright (C) 2016 CERN.
```
Comment (ppanero) [inline datacite/schema42.py]: 2019

File: tests/test_client.py [hunk 4]
```diff
@@ -58,19 +58,20 @@ def test_ssl_error(requests):
...
-    httpretty.register_uri(
-        httpretty.GET,
-        "{0}doi/10.1234/1".format(APIURL),
-        body=callback,
-    )
-
-    d = get_client(timeout=0.1)
-    with pytest.raises(DataCiteHttpError):
-        d.doi_get("10.1234/1")
+# Haven't gotten timeout to work correctly with responses
```
Comment (ppanero) [inline tests/test_client.py]: Not fully sure of its relation. But it seems timeouts are not supported in responses? https://github.com/getsentry/responses/pull/115

File: setup.py [hunk 5]
```diff
@@ -20,15 +20,15 @@
 tests_require = [
     'check-manifest>=0.25',
     'coverage>=4.0',
-    'httpretty>=0.8.14',
+    'responses>=0.10.6',
```
Comment (ppanero) [inline setup.py]: @lnielsen Do you see any strong issue that ``httpretty`` to ``responses`` might imply? License is Apache, seems to be well maintained, use is almost 1-1 match...

File: datacite/schema42.py (hunk 3 above)
Comment (tmorrell) [reply datacite/schema42.py]: I added Caltech and left in CERN (alot is the same as schema41)

File: setup.py (hunk 5 above)
Comment (lnielsen) [reply setup.py]: No problem with Apache license.....only GPL and AGPL are an issue.

**Comment by lnielsen** [comment]
I'm in process of merging the PR. However, there's some strange things with the official JSON schema.

I find it odd that ``agency`` (and possibly ``schemaVersion``) are required properties. I will have a closer look tomorrow.

**Comment by tmorrell** [comment]
I've removed agency as a requires property and updated my pull request with DataCite (https://github.com/datacite/schema/pull/55).  You could argue that schemaVersion maps to the XML which includes the schema, but we could remove that as well.  

The DataCite "official" JSON version of the schema is still a bit of a work in progress, so if there are other things we want to change it's a good time.  I've just done the minimal modifications to get everything to work.

**Comment by lnielsen** [comment]
🚀 Merged :-) There's still a couple of issues to discuss:

- Publisher lang attribute (datacite/schema#60)
- Changes in datacite/schema#59

Also, I'd like to discuss the use of ``uniqueItems`` and why ``types.resourceType`` is required, but I'll open separate issues on the datacite/schema repository for those, as the discussion is better suited there.

**Pull request merged**